### PR TITLE
[Bugfix] Add LightningOptimizer parity test and resolve AMP bug

### LINKED ISSUE #5159: performance loss from 1.0.8 to 1.1.* when using 16 bit precision
## 🐛 Bug

After updating pytorch-lightning from 1.0.8 to 1.1.0/1.1.1 the use of 16 bit precision destroys the performances.
In my actual code of object detection losses are by a factor of 4 larger at the beginning than compared to 32 bit or 16 bit with pl 1.08.
They converge to a much higher value and the resulting model lost its detection capabilities completely.
To replicate I tested the pl notebooks and the `06-cifar10-baseline.ipynb` also shows this and the classification accuracy corresponds to guessing the class when switching from 32 to 16 bit.
I integrated it into the BoringModel notebook and the problem is also happening in google colab.

## Please reproduce using [the BoringModel and post here](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing)

https://colab.research.google.com/drive/1FqXG9Xw9gVZxnwiGnjsHpAtb-vUqFaob?usp=sharing

### Expected behavior

Same performance for 32 and 16 bit.

### Environment

* CUDA:
	- GPU:
		- Tesla P100-PCIE-16GB
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.18.5
	- pyTorch_debug:     True
	- pyTorch_version:   1.7.0+cu101
	- pytorch-lightning: 1.1.1
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### LINKED ISSUE #5165: Reproducibility bug between LightningOptimizer activate / deactivate
## Please reproduce using [the BoringModel and post here](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing)

```
import os
import torch
from torch.utils.data import Dataset
from pytorch_lightning import Trainer, LightningModule, seed_everything


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        """
        Testing PL Module

        Use as follows:
        - subclass
        - modify the behavior for what you want

        class TestModel(BaseTestModel):
            def training_step(...):
                # do your own thing

        or:

        model = BaseTestModel()
        model.training_epoch_end = None

        """
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def loss(self, batch, prediction):
        # An arbitrary loss to have a loss that updates the model weights during `Trainer.fit` calls
        return torch.nn.functional.mse_loss(prediction, torch.ones_like(prediction))

    def step(self, x):
        x = self.layer(x)
        out = torch.nn.functional.mse_loss(x, torch.ones_like(x))
        return out

    def training_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        print(loss)
        return {"loss": loss}

    def training_step_end(self, training_step_outputs):
        return training_step_outputs

    def training_epoch_end(self, outputs) -> None:
        torch.stack([x["loss"] for x in outputs]).mean()

    def validation_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"x": loss}

    def validation_epoch_end(self, outputs) -> None:
        torch.stack([x['x'] for x in outputs]).mean()

    def test_step(self, batch, batch_idx):
        output = self.layer(batch)
        loss = self.loss(batch, output)
        return {"y": loss}

    def test_epoch_end(self, outputs) -> None:
        torch.stack([x["y"] for x in outputs]).mean()

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]


#  NOTE: If you are using a cmd line to run your script,
#  provide the cmd line as below.
#  opt = "--max_epochs 1 --limit_train_batches 1".split(" ")
#  parser = ArgumentParser()
#  args = parser.parse_args(opt)

def run_test():
    seed_everything(42)

    # fake data
    train_data = torch.utils.data.DataLoader(RandomDataset(32, 64))
    val_data = torch.utils.data.DataLoader(RandomDataset(32, 64))

    # model
    before = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        max_epochs=1,
        limit_train_batches=4,
        limit_val_batches=0,
        weights_summary=None,
        gpus=1,
        precision=16,
    )
    trainer.fit(before, train_data, val_data)

    seed_everything(42)
    # fake data
    train_data = torch.utils.data.DataLoader(RandomDataset(32, 64))
    val_data = torch.utils.data.DataLoader(RandomDataset(32, 64))

    # model
    after = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        max_epochs=1,
        limit_train_batches=4,
        limit_val_batches=0,
        weights_summary=None,
        gpus=1,
        precision=16,
        enable_pl_optimizer=False
    )
    trainer.fit(after, train_data, val_data)

    # Assert model parameters are identical after fit
    for before, after in zip(before.parameters(), after.parameters()):
        assert torch.equal(before, after), 'Model parameters are different'


if __name__ == '__main__':
    run_test()
```
Credit to @SeanNaren 

### Environment

**Note**: `Bugs with code` are solved faster ! `Colab Notebook` should be made `public` !

* `IDE`: Please, use our python [bug_report_model.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
) template.

* `Colab Notebook`: Please copy and paste the output from our [environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py) (or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #5165
Fixes #5159

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -489,6 +489,10 @@ def optimizer_step(self, optimizer, opt_idx, batch_idx, train_step_and_backward_
                 'native PyTorch amp and lbfgs are not compatible.'
                 ' To request, please file a Github issue in PyTorch and tag @mcarilli')
 
+        if not isinstance(optimizer, LightningOptimizer):
+            # wraps into LightingOptimizer only for running step
+            optimizer = LightningOptimizer.to_lightning_optimizer(optimizer, self.trainer)
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/training_loop.py:494]: Just something to consider, this wraps the optimizer but never unwraps right?

File: tests/trainer/optimization/test_parity_automatic_optimization.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+@pytest.mark.parametrize(["precision", "amp_backend", "gpus"], [
+    pytest.param(32, "native", 0),
+])
+@pytest.mark.parametrize('accumulate_grad_batches', [1])
+def test_lightning_optimizer_and_no_lightning_optimizer_equality_check_optim_calls(
+        tmpdir,
+        precision,
+        amp_backend,
+        gpus,
+        accumulate_grad_batches):
```
Comment (Borda) [inline tests/trainer/optimization/test_parity_automatic_optimization.py:73]:
```suggestion
        accumulate_grad_batches
    ):
```

File: tests/trainer/optimization/test_parity_automatic_optimization.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+        assert mock_sgd_zero_grad.call_count == (expected_num_batches // accumulate_grad_batches)
+        assert mock_sgd_step.call_count == mock_adam_step.call_count
+        assert mock_sgd_step.call_count == mock_adam_step.call_count
+        assert mock_sgd_zero_grad.call_count == mock_adam_zero_grad.call_count
+        assert mock_sgd_zero_grad.call_count == mock_adamw_zero_grad.call_count
+
+
+def run_lightning_optimizer_equality(
+        optimizer_is_mocked=False,
+        **trainer_kwargs):
```
Comment (Borda) [inline tests/trainer/optimization/test_parity_automatic_optimization.py:106]:
```suggestion
        **trainer_kwargs
    ):
```
easier to see where a func declaration ends

**Review by Borda — COMMENTED** [review]
is there a simple way how to unify the two files, reuse function, I have a feeling that they are almost the same

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py:494]: It wraps it for `optimizer_step` only but we don't assigned it back to the Trainer, so `self.optimizers` would always be the unwraped ones. Am I missing something ?

**Comment by tchaton** [comment]
> is there a simple way how to unify the two files, reuse function, I have a feeling that they are almost the same

You are right, I re-used code between tests.

File: tests/trainer/optimization/test_parity_automatic_optimization.py [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+from tests.base.boring_model import BoringModel
+
+"""
+TODO:
+For both automatic / manual optimization
+    - Test dp, ddp, ddp2
+    - Apex
+    - Random accumulated_grad_batches (bug)
+    - Multiple optimizers
+"""
```
Comment (Borda) [inline tests/trainer/optimization/test_parity_automatic_optimization.py:35]:
```suggestion
# TODO:
# For both automatic / manual optimization
#     - Test dp, ddp, ddp2
#     - Apex
#     - Random accumulated_grad_batches (bug)
#     - Multiple optimizers
```

File: tests/trainer/optimization/test_parity_automatic_optimization.py [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+    pytest.param(32, "native", 0),
+    pytest.param(16, "native", 1, marks=pytest.mark.skipif(not torch.cuda.is_available(), reason='Requires GPU')),
+])
+@pytest.mark.parametrize('accumulate_grad_batches', [1, 7])
+def test_lightning_optimizer_and_no_lightning_optimizer_equality(
+        tmpdir,
+        precision,
+        amp_backend,
+        gpus,
+        accumulate_grad_batches):
```
Comment (Borda) [inline tests/trainer/optimization/test_parity_automatic_optimization.py:177]:
```suggestion
        accumulate_grad_batches,
):
```

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/core/lightning.py [hunk 6]
```diff
@@ -1126,6 +1126,7 @@ def training_step(...):
 
         # backward
         self._running_manual_backward = True
+        loss /= self.trainer.accumulate_grad_batches
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:1129]: I don't understand this addition. Can you explain?
Why does it not break any existing tests?

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:494]: @tchaton you can move the if statement inside the `to_lightning_optimizer` function to achieve idempotence and avoid code duplication.

**Review by williamFalcon — APPROVED** [review]

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (SeanNaren) [reply pytorch_lightning/trainer/training_loop.py:494]: My mistake! You're right

File: pytorch_lightning/core/lightning.py (hunk 6 above)
Comment (SeanNaren) [reply pytorch_lightning/core/lightning.py:1129]: Also making an assumption that we should average across accumulated batches right?

Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1129]: We have the same assumption for `automatic optimization`.  I added this line to uniformize behaviours but the mean assumption makes more sense to me than the sum.

Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:1129]: ok but does this have to be in the lightningmodule? Wouldn't this go to the training loop where all the other accumulation logic is?

Comment (tchaton) [reply pytorch_lightning/core/lightning.py]: Done, moved it internally.

**Review by awaelchli — APPROVED** [review]
the tests are a bit complex, I don't really understand them but I was able to verify that they fail on the master branch with the wrong loss values so I suppose the scaling bug is fixed 👍

**Comment by tchaton** [comment]
> the tests are a bit complex, I don't really understand them but I was able to verify that they fail on the master branch with the wrong loss values so I suppose the scaling bug is fixed 👍

The test are performing parity comparison with vanilla training in several scenarios and making sure everything matches properly when using enable_pl_optimizer=True/False. We will need to add more parity test such as Apex, DDP modes, multi optimisers.

Best,
T.C

**Review by SeanNaren — APPROVED** [review]
